File: infinite_sense_core/src/ptp.cpp
#include "ptp.h"
#include "usb.h"
#include "log.h"
#include <thread>
#include <chrono>

namespace infinite_sense {

constexpr char func_name[] = "f";
constexpr char func_type_a[] = "a";
constexpr char func_type_b[] = "b";

void Ptp::SetUsbPtr(const std::shared_ptr<serial::Serial>& serial_ptr) { serial_ptr_ = serial_ptr; }

void Ptp::SetNetPtr(const std::shared_ptr<UDPSocket>& net_ptr, const std::string& target_ip,
                    const unsigned short port) {
  net_ptr_ = net_ptr;
  target_ip_ = target_ip;
  port_ = port;
}

void Ptp::ReceivePtpData(const nlohmann::json& data) {
  try {
    const std::string func = data.at(func_name);

    if (func == func_type_a) {
      HandleTimeSyncRequest(data);
    } else if (func == func_type_b) {
      HandleTimeSyncResponse(data);
    }
  } catch (const nlohmann::json::exception& e) {
    LOG(ERROR) << "PTP JSON parse error: " << e.what();
  }
}

void Ptp::HandleTimeSyncRequest(const nlohmann::json& data) {
  try {
    time_t1_ = data.at(func_type_a);
    time_t2_ = data.at(func_type_b);
    updated_t1_t2_ = true;
  } catch (const nlohmann::json::exception& e) {
    LOG(ERROR) << "Invalid 'a' message format: " << e.what();
  }
}

void Ptp::HandleTimeSyncResponse(const nlohmann::json& data) {
  try {
    const uint64_t t3 = data.at(func_type_a);
    const uint64_t t4 = GetCurrentTimeUs();

    if (updated_t1_t2_) {
      const int64_t delay = static_cast<int64_t>(t4 - t3 + time_t2_ - time_t1_) / 2;
      const int64_t offset = static_cast<int64_t>(time_t2_ - time_t1_ - t4 + t3) / 2;

      const nlohmann::json response = {
          {func_name, func_type_b},
          {func_type_a, delay},
          {func_type_b, offset},
      };

      SendJson(response);
      updated_t1_t2_ = false;
    }
  } catch (const nlohmann::json::exception& e) {
    LOG(ERROR) << "Invalid 'b' message format: " << e.what();
  }
}

void Ptp::SendPtpData() const {
  const uint64_t mark = GetCurrentTimeUs();

  const nlohmann::json data = {
      {func_name, func_type_a},
      {func_type_a, mark},
  };

  SendJson(data);
  std::this_thread::sleep_for(std::chrono::milliseconds(100));
}

void Ptp::SendJson(const nlohmann::json& data) const {
  const std::string out = data.dump() + "\n";

  if (net_ptr_) {
    net_ptr_->sendTo(out.data(), out.size(), target_ip_, port_);
  } else if (serial_ptr_) {
    serial_ptr_->write(reinterpret_cast<const uint8_t*>(out.data()), out.size());
  } else {
    LOG(WARNING) << "No valid output interface for PTP message.";
  }
}

uint64_t Ptp::GetCurrentTimeUs() {
  return std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::system_clock::now().time_since_epoch())
      .count();
}

}  // namespace infinite_sense

File: python/infinite_sense_core/ptp.py
import json

from .times import get_current_time_us

class Ptp:
    FUNC_NAME = "f"
    FUNC_TYPE_A = "a"
    FUNC_TYPE_B = "b"

    def __init__(self):
        self.serial_ptr = None
        self.net_ptr = None
        self.target_ip = None
        self.port = None

        self.time_t1 = 0
        self.time_t2 = 0
        self.updated_t1_t2 = False

    def set_usb_ptr(self, serial_ptr):
        self.serial_ptr = serial_ptr

    def set_net_ptr(self, net_ptr, target_ip: str, port: int):
        self.net_ptr = net_ptr
        self.target_ip = target_ip
        self.port = port

    def receive_ptp_data(self, data: dict):
        try:
            func = data[self.FUNC_NAME]

            if func == self.FUNC_TYPE_A:
                self.handle_time_sync_request(data)
            elif func == self.FUNC_TYPE_B:
                self.handle_time_sync_response(data)
        except Exception as e:
            print(f"[ERROR] PTP JSON parse error: {e}")

    def handle_time_sync_request(self, data: dict):
        try:
            self.time_t1 = data[self.FUNC_TYPE_A]
            self.time_t2 = data[self.FUNC_TYPE_B]
            self.updated_t1_t2 = True
        except Exception as e:
            print(f"[ERROR] Invalid 'a' message format: {e}")

    def handle_time_sync_response(self, data: dict):
        try:
            t3 = data[self.FUNC_TYPE_A]
            t4 = get_current_time_us()

            if self.updated_t1_t2:
                delay = int((t4 - t3 + self.time_t2 - self.time_t1) / 2)
                offset = int((self.time_t2 - self.time_t1 - t4 + t3) / 2)

                response = {
                    self.FUNC_NAME: self.FUNC_TYPE_B,
                    self.FUNC_TYPE_A: delay,
                    self.FUNC_TYPE_B: offset
                }

                self.send_json(response)
                self.updated_t1_t2 = False
        except Exception as e:
            print(f"[ERROR] Invalid 'b' message format: {e}")

    def send_ptp_data(self):
        mark = get_current_time_us()

        data = {
            self.FUNC_NAME: self.FUNC_TYPE_A,
            self.FUNC_TYPE_A: mark
        }
        self.send_json(data)

    def send_json(self, data: dict):
        out = json.dumps(data) + "\n"
        if self.net_ptr:
            self.net_ptr.sendto(out.encode(), (self.target_ip, self.port))
        elif self.serial_ptr:
            self.serial_ptr.write(out.encode())
        else:
            print("[WARNING] No valid output interface for PTP message.")
